### ISSUE TITLE
FastForward and/or ContextShift do not work with Qwen3-Next-80B-A3B-Thinking

### ISSUE THREAD

**Issue by NGgmnBQGwR** [LostRuins/koboldcpp#1895]
Using v1.104 and Qwen3-Next-80B-A3B-Thinking-UD-Q4_K_XL.gguf from https://huggingface.co/unsloth/Qwen3-Next-80B-A3B-Thinking-GGUF.

When using 'Retry' to re-generate model's reply, some (all?) of context is reprocessed every time.
For example, model wrote a page of text, but I didn't like the last few sentences. I remove them, click 'Generate More' (with enabled '[X] Continue Bot Reply').
> Processing Prompt [BATCH] (1796 / 1796 tokens)

Model generates new text. I don't like it either, so I click 'Retry'.
> Processing Prompt [BATCH] (512 / 1796 tokens)

I stop generation and click 'Retry' again.
> Processing Prompt [BATCH] (512 / 1796 tokens)

Same with regular chat - the Y in 'Processing Prompt X/Y' is increasing with every message (i.e. 1300 for first message, 1600 for second, 2100 for third one).
Tried different thinking presets (Normal/Forced) and with 'Send Only Latest COT' vs 'Send All COTs' - no difference, so I don't think it's related to the thinking part.

I've been using Mistral-Small-3.2-24B-Instruct-2506-UD-Q6_K_XL.gguf for a long time and never had this issue with it.

I'm inclined to think that it's something on my end, since I haven't seen anyone else with this problem, but how can I troubleshoot this?

Launch parameters:
`koboldcpp-nocuda.exe --highpriority --launch --nommap --multiplayer --threads 7 --usevulkan 0 --gpulayers 6 --contextsize 16384 Qwen3-Next-80B-A3B-Thinking-UD-Q4_K_XL.gguf`

System:
W10, 64GB RAM (I still have ~8GB of RAM free after launching the model), 8GB VRAM.

**Comment by LostRuins** [comment]
Qwen3next is a hybrid model. That means that normal fast forward won't work. Instead, please launch with --smartcache or turn it on in the tokens tab. That will enable checkpointing for this model

**Comment by NGgmnBQGwR** [comment]
Thank you for the quick reply!

But I can't figure out what's wrong - everything (retrying + new messages) still re-process context every time.

Startup, click 'Generate More'. Model generates text, and I abort it in the middle:
<details>
Processing Prompt [BATCH] (1020 / 1020 tokens)
Generating (195 / 768 tokens)state_write_data: writing state
state_write_data: - writing model info
state_write_data: - writing output ids
state_write_data: - writing logits
state_write_data: - writing embeddings
state_write_data: - writing memory module
state_write_data: writing state
state_write_data: - writing model info
state_write_data: - writing output ids
state_write_data: - writing logits
state_write_data: - writing embeddings
state_write_data: - writing memory module

Generation Aborted

KV Save State 4: Created SaveState of 1215 tokens, costing 104 MB.
</details>

I click 'Retry', then abort it:

<details>
[SmartCache RNN No Match, Already saved in slot 4]

Processing Prompt [BATCH] (1202 / 1202 tokens)
Generating (1 / 768 tokens)
(EOS token triggered! ID:151645)state_write_data: writing state
state_write_data: - writing model info
state_write_data: - writing output ids
state_write_data: - writing logits
state_write_data: - writing embeddings
state_write_data: - writing memory module
state_write_data: writing state
state_write_data: - writing model info
state_write_data: - writing output ids
state_write_data: - writing logits
state_write_data: - writing embeddings
state_write_data: - writing memory module

KV Save State 3: Created SaveState of 1203 tokens, costing 104 MB.
</details>

'Retry' and abort again:

<details>
[SmartCache RNN No Match, Already saved in slot 3]

Processing Prompt [BATCH] (1202 / 1202 tokens)
Generating (34 / 768 tokens)state_write_data: writing state
state_write_data: - writing model info
state_write_data: - writing output ids
state_write_data: - writing logits
state_write_data: - writing embeddings
state_write_data: - writing memory module

Generation Aborted
state_write_data: writing state
state_write_data: - writing model info
state_write_data: - writing output ids
state_write_data: - writing logits
state_write_data: - writing embeddings
state_write_data: - writing memory module

KV Save State 2: Created SaveState of 1236 tokens, costing 104 MB.
</details>

Reply + abort:

<details>
[SmartCache RNN No Match, Already saved in slot 1]

Processing Prompt [BATCH] (1358 / 1358 tokens)
Generating (151 / 768 tokens)state_write_data: writing state
state_write_data: - writing model info
state_write_data: - writing output ids
state_write_data: - writing logits
state_write_data: - writing embeddings
state_write_data: - writing memory module
state_write_data: writing state
state_write_data: - writing model info
state_write_data: - writing output ids
state_write_data: - writing logits
state_write_data: - writing embeddings
state_write_data: - writing memory module

Generation Aborted

KV Save State 0: Created SaveState of 1509 tokens, costing 111 MB.
</details>

Another reply + abort:

<details>
[SmartCache RNN No Match, Already saved in slot 0]

Processing Prompt [BATCH] (1525 / 1525 tokens)
Generating (22 / 768 tokens)
Generation Aborted
state_write_data: writing state
state_write_data: - writing model info
state_write_data: - writing output ids
state_write_data: - writing logits
state_write_data: - writing embeddings
state_write_data: - writing memory module
state_write_data: writing state
state_write_data: - writing model info
state_write_data: - writing output ids
state_write_data: - writing logits
state_write_data: - writing embeddings
state_write_data: - writing memory module

KV Save State 4: Created SaveState of 1547 tokens, costing 112 MB.
</details>

**Comment by LostRuins** [comment]
Can you try with a shorter prompt? second prompt should contain the same starting part as first prompt. To be clear, the smart cache state save/load only works for fast forwarding a prompt, not for shifting

**Comment by simadude** [comment]
I believe I have the same issue with Nemotron-3-Nano-30B-A3B (also hybrid model to my knowledge). Kobold.cpp keeps reprocessing the whole prompt, while llama.cpp doesn't.

When I keep requesting for completion with the same prompt (with smart cache enabled), it shows me `[SmartCache RNN No Match, Already saved in slot 4]`. Then after each consequent request the slot number changes to 3, 2, 1, 0 and back to 4.

I have tried with prompt lengths of 1008 and 398, both yielding the same result, with and without smart cache.

**Comment by LostRuins** [comment]
Are you sure the payloads you submitted have a common prefix to fast forward? The reprocessing occurs when it cannot find a match from the previous prompt.

Try this simple test: I have loaded Nemotron-3-Nano-30B-A3B-Q2_K_L.gguf with `--smartcache` enabled
Start a new blank instruction session in koboldai lite. 
Send the prompt "tell me a joke"
Let the AI generate the joke. For me, it took 65 tokens total for me, observe that it saved after generation was completed.

<img width="1145" height="410" alt="Image" src="https://github.com/user-attachments/assets/4798e8c7-2370-47d3-82c0-935ab5f314af" />

<img width="708" height="485" alt="Image" src="https://github.com/user-attachments/assets/5b13feab-6f86-4dc2-89d6-cb32b149748f" />

Now send a follow up prompt, "`another one please`". It should continue with the existing prompt (without reprocessing much)

<img width="1196" height="521" alt="Image" src="https://github.com/user-attachments/assets/d17974f4-8c98-457b-8f65-5560bd233c7b" />

Now press "Retry". It should regenerate the second turn using the first turn's checkpoint, with minimal reprocessing.

<img width="1171" height="838" alt="Image" src="https://github.com/user-attachments/assets/84ca401f-35e9-4347-bd24-548125df4a98" />

Notice it loaded and reused the original 65 tokens, showing that the cache is working.

**Comment by simadude** [comment]
So I have tested a bit more and it turns out that it reprocesses prompts only when you try to regenerate the message. Like if I append more context to already generated message then it's fine, but if I try to regenerate then it reprocesses the whole prompt.

**Comment by simadude** [comment]
Okay, maybe I haven't tested enough, since SOMETIMES it doesn't reprocess the whole context.

Anyway, here's what I used to trigger reprocessing every time:
```
curl -d '{"prompt": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla lacinia nibh a dignissim fermentum. Integer non massa sed metus consectetur rutrum.", "max_tokens": 20}' http://localhost:5001/v1/completions
```

**Comment by LostRuins** [comment]
Well that makes sense, because the snapshot was taken *after* the generation completed (so position 34 + 20 = 54)

The second gen sends the original 34 tokens, there is no snapshot as pos 34.

I guess I could snapshot both positions (before generating and after generating) to make it more robust.

**Comment by NGgmnBQGwR** [comment]
I think my issue was fixed by using 'Send All CoTs' from the beginning. There's still some reprocessing (mainly when model ran out of 768 allocated tokens and I click 'Generate More' - it loads from cache, but then processes ~776 tokens), but now I see cache being used, and follow-up replies no longer process everything.

Sorry for the confusion and thanks for the great feature!
